src/snowflake/snowpark/_internal/ast/batch.py

```diff
@@ -95,55 +124,142 @@ def eval(self, target: proto.Bind) -> None:
         Args:
             target: The variable to evaluate.
         """
-        stmt = self._request.body.add()
-        stmt.eval.bind_id = target.uid
+        with self._lock:
+            stmt = proto.Stmt()
+            stmt.eval.bind_id = target.uid
+            self._cur_stmts.append(stmt)
 
-    def flush(self) -> SerializedBatch:
-        """Ties off a batch and starts a new one. Returns the tied-off batch."""
-        req_id: str = str(self._request_id)
+    def create_full_ast_request(
+        self,
+    ) -> proto.Request:
+        """Create fully contained AST request with all dependent AST objects."""
+        if not self._cur_stmts:
+            return proto.Request()
+
+        def get_dependent_bind_ids(ast: Any) -> set[int]:
+            """Retrieve the dependent AST IDs required for this AST object."""
+            dependent_ids = set()
+            if isinstance(ast, Iterable) and not isinstance(ast, str):
+                for c in ast:
+                    dependent_ids.update(get_dependent_bind_ids(c))
+            elif hasattr(ast, "DESCRIPTOR"):
+                descriptor = ast.DESCRIPTOR
+                if descriptor.name in (
+                    "DataframeRef",
+                    "RelationalGroupedDataframeRef",
+                    "FnIdRefExpr",
+                    "FnRef",
+                    "IndirectTableFnIdRef",
+                ):
+                    dependent_ids.add(ast.id)  # type: ignore[union-attr]
+                else:
+                    for f in descriptor.fields:
+                        c = getattr(ast, f.name)
+                        if (
+                            isinstance(c, Iterable)
+                            and not isinstance(c, str)
+                            and len(c) > 0  # type: ignore[arg-type]
+                        ) or (isinstance(c, Message) and c.ByteSize() > 0):
+                            dependent_ids.update(get_dependent_bind_ids(c))
+            return dependent_ids
+
+        # Tracks the eval statements by their bind ID for retrieval later.
+        eval_stmt_bind_ids: dict[int, proto.Stmt] = dict()
+        # Collect all bind statements in the current set of statements.
+        cur_request_bind_ids = []
+        for stmt in self._cur_stmts:
+            if stmt.eval.bind_id != 0:
+                eval_stmt_bind_ids[
+                    stmt.eval.bind_id
+                ] = stmt  # stores the eval statement
+            else:
+                self._prev_bind_stmts[stmt.bind.uid] = stmt  # stores the bind statement
+                cur_request_bind_ids.append(stmt.bind.uid)
+
+        # Keep track of all the bind IDs (including dependencies) that we need to send over.
```

Comment:
this function is very long, can you split up into parts? similar to ASTBuilder design?